File: 刷题/牛客网/腾讯/贪吃的小Q/tanchi.cpp
#include <stdio.h>
#include <math.h>//数学函数库
// double ceil(double )//向上取整函数原型

int BinarySearch(int low, int high, int n, int m) {
 int max, mid, next, sum;
 float midTmp, nextTmp;
 while(low<=high){
  midTmp = ((float)(low+high))/2;
  mid = (int)ceil(midTmp);
  next = mid;
        sum = 0;
  for(int i=0; i<n ;i++) {
   sum += next;
   nextTmp = ((float)next)/2;
   next = (int)ceil(nextTmp);
   if(next==1) {
    sum += (n-i-1);
    break;
   }
  }
  if(sum>m) {
   high = mid;
  } else if(sum<m) {
   low = mid;
  } else if(sum==m) {
   max = mid;
   break;
  }
 }
 return max;
}
// int BinarySearch(int low, int high, int n, int m) {
//  int mid = (int)ceil((low+high)/2);
//  int next = mid, sum = 0, max;
//  for(int i=0; i<n ;i++) {
//   sum += next;
//   next = (int)ceil(next/2);
//  }
//  if(sum>m) {
//   max = BinarySearch(low, mid, n, m);
//  } else if(sum<m) {
//   max = BinarySearch(mid, high, n, m);
//  } else if(sum==m) {
//   max = mid;
//  }
//  return max;
// }
int main() {
 int n,m;
 scanf("%d %d", &n, &m);
 int ceil = m-n+1;
 int res = BinarySearch(1, ceil, n, m);
 printf("%d", res);
 return 0;
}

File: 刷题/牛客网/腾讯/贪吃的小Q/tanchi.js
const readline = require('readline');
const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
});

rl.on('line', line => {
    let tokens = line.split(' ');
    let n = parseInt(tokens[0]);
    let m = parseInt(tokens[1]);
    // n：天数，m：巧克力数
    console.log(binarySearch(n, m));
});

function binarySearch(n, m) {
    // 假设后面几天都是一块巧克力，得到第一天的数量取值范围为[1, m-n]
    // 然后在[1, m-n]中取最合适的数字，使得总和不超过m
    if (n == 1)
        return m;
    let _m = m - n + 1;
    let p = 1, q = _m;
    let first = _m;
    while (p <= q) {
        let mid = Math.ceil((p + q) / 2);
        let next = mid;
        let sum = 0;
        for (let i = 0; i < n; i++) {
            sum += next;
            next = Math.ceil(next / 2);
            if (next == 1) {
                sum += n - i - 1;
                break;
            }
        }
        if (sum > m && q != mid) {
            // 估值高了，往下
            q = mid - 1;
        } else if (sum < m && p != mid) {
            // 估值低了，往上
            p = mid + 1;
        } else {
            first = mid;
            if (q == mid) {
                first--;
            }
            if (p == mid) {
                first++;
            }
            break;
        }
    }
    return first;
}
